dtlib: error out on duplicate node names

mbolivar-nordic opened the pull request:
This fix for a medium priority DT bug requires tree-wide cleanup.

Attempts to define two nodes with the same name within a single set of curly brackets should fail.

For example, this is invalid DTS according to dtc:

```dts
    / { foo {}; foo {}; };
```

By contrast, this is valid since the node named 'foo' appears twice in two different sets of curly brackets:

```dts
    / { foo {}; };
    / { foo {}; };
```

Zephyr's dtlib currently does not error out on the invalid condition. Before fixing that, we have to apply patches with updated devicetrees throughout zephyr and the STM32 HAL that remove duplicate nodes. Then we can make dtlib match dtc and add a regression test.

Fixes: #49590

mbolivar-nordic committed:
dtlib: error out on duplicate node names

Attempts to define two nodes with the same name within a single set of
curly brackets should fail.

For example, this is invalid DTS according to dtc:

    / { foo {}; foo {}; };

By contrast, this is valid since the node named 'foo' appears twice in
two different sets of curly brackets:

    / { foo {}; };
    / { foo {}; };

Zephyr's dtlib currently does not error out on the invalid condition.

Now that Zephyr itself has been updated to not include such nodes (to
the best of my ability), we can fix this divergence from current dtc
behavior and add a regression test in dtlib.

Fixes: #49590
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
I knew it was too much to hope for that we wouldn't have a bunch of new failures :).

mbolivar-nordic committed:
dts: nxp: rt11xx: fix lpuart12 base addresses

The correct address is 0x40c28000 on these SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: intel: cavs25: hack around awkward DT representation

Apparently, downstream drivers are allocating devices using
DT_NODELABEL(alh0) and DT_NODELABEL(alh1) on a node with compatible
intel,alh-dai. These represent something like "channels" within the
device.

This is a strange choice. It would be better to do something like have
a property with the count of "channels", or one child node per
channel.

It is also a dtc error to have duplicate node names like this, and
will be an error in Zephyr's dtlib soon.

For now, work around this representation issue by replacing it with
something equivalent that doesn't have duplicate nodes. The
representation should be addressed at some point, but not by me.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
boards: hifive1: fix PWM pinctrl

A likely copy/paste error is making the pwm2_1_default pinmux setting
override the value for pwm1_1_default. Fix it.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
boards: it8xxx2_evb: fix invalid DTS

The duplicated aliases node in this board's DTS file is an error when
compiling it with dtc. Zephyr's dtlib doesn't report the error yet,
but it will soon, so rearrange the alias definitions under a single
node. No functional changes expected.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: mec172x: pinctrl fixes

Two issues:

- The nsmi_gpio107 node is being defined twice with the same pinmux
  property value. This is an error when compiling the file with dtc.
  Zephyr's dtlib doesn't currently error out on this, but it will soon.

  Fix this by removing one of the redundant definitions.

- The eeprom_clk_gpio117 node label is referring to a node named
  gpspi_clk_gpio117, which is already defined in the same file, but
  with a different pinmux property value. This looks like a clear
  copy/paste error causing invalid pinmux settings for the
  gpspi_clk_gpio117 node.

  Fix it by aligning the node name with the node label to create
  a separate node with its own pinmux value.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
boards: arduino_nicla_sense_me: pinctrl fixes

Two pinctrl node labels for this board are referring to nodes that are
already defined higher in the file with different pinmux property
values.

This looks like a clear copy/paste error. Fix it by renaming the nodes
to create two distinct nodes in each case instead of overriding the
previous node definition. This will also make the file compile in dtc,
which regards duplicate node definitions as errors, and will be a
required fix for when Zephyr's dtlib does the same.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
west.yml: update stm32 HAL

Bring in updated pinctrl files. These include fixes which remove
duplicate nodes.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

dcpleung commented on dts/xtensa/intel/intel_adsp_cavs25.dtsi:
```diff
@@ -260,11 +260,13 @@
 			status = "okay";
 		};
 
+#if 0		   /* FIXME what's the right address for this node? */
```
IIRC, it is actually the same IP block. There are multiple "instances" or "channels" where they share the some common registers (e.g. control and status). That's why they have the same base address. They appear as two separate devices because these will be used by downstream drivers (referred using the `DT_*` macros).

Not sure what would be a good way to express this in devicetree.

dcpleung replied:
@mbolivar-nordic Thanks.

@abonislawski @juimonen Could you take a look and update the drivers and device trees?

nandojve reviewed (changes requested):
For SAMs SoC: HI @mbolivar-nordic  I can be wrong but I think if you  rename it to group2 it may be enough. In this case, there is no need to exclude first group, otherwise it will drop RX configuration. This could be made on all board occurrences.

erwango commented on west.yml:
```diff
@@ -129,7 +129,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 5b1610c3cf46589d50e2634ba240b7b38bc59c53
+      revision: pull/149/head
```
Please use https://github.com/zephyrproject-rtos/hal_stm32/pull/150

mbolivar-nordic committed:
boards: atmel SAM: pinctrl fixes

Various boards are incorrectly specifying two pin control groups named
'group1' in the same parent node. Fix this by splitting them into two
groups, replacing redundant 'group1' with 'group2'.

Suggested-by: Gerson Fernando Budke <nandojve@gmail.com>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@nandojve @erwango comments addressed, thanks for your help

mbolivar-nordic commented:
The remaining CI error seems to be due to an issue in the updated STM32 HAL PR.

gmarull commented:
> The remaining CI error seems to be due to an issue in the updated STM32 HAL PR.

Seems to be fixed with latest changes in the HAL PR.

mbolivar-nordic commented:
> Seems to be fixed with latest changes in the HAL PR.

I merged that and updated the manifest, then. Thanks.
